tools/hanabi_moves.py

```diff
@@ -1,94 +1,97 @@
-class HanabiCardAction(object):
-    def __init__(self, player_id, card):
+import abc
+
+class HanabiAction(object):
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, player_id):
         self.player_id = player_id
+
+    @abc.abstractmethod
+    def is_valid(self, game_info):
+        pass
+
+    @abc.abstractmethod
+    def execute(self, table):
+        pass
+
+class HanabiCardAction(HanabiAction):
+    def __init__(self, player_id, card):
+        super(HanabiCardAction, self).__init__(player_id)
         self.card = card
+        self.action_type = ''
 
     def __str__(self):
-        return "Player {id} {action} {card}".format(id = self.player_id, action = self.action(), card = str(self.card))
-
-    @classmethod
-    def action(self):
-        return "N/A"
+        return "Player {id} {action} {card}".format(id = self.player_id, action = self.action_type, card = str(self.card))
 
 class HanabiPlayAction(HanabiCardAction):
-    
+
     def __init__(self, player_id, card):
         super(HanabiPlayAction, self).__init__(player_id, card)
+        self.action_type = 'played'
 
-    @classmethod
-    def parse_move(self, move):
-        pass
+    def is_valid(self, game_info):
+        return True
 
-    @classmethod
-    def action(self):
-        return "played"
-
-    @staticmethod
-    def can_parse_move(move):
-        return (all(key in move for key in ("play_type", "card")) and
-            move["play_type"] is "play") 
+    def execute(self, table):
+        table.play_card(self.player_id, self.card)
 
 class HanabiDiscardAction(HanabiCardAction):
     def __init__(self, player_id, card):
         super(HanabiDiscardAction, self).__init__(player_id, card)
+        self.action_type = 'discarded'
 
-    @classmethod
-    def action(self):
-        return "discarded"
 
-    @staticmethod
-    def can_parse_move(move):
-        return (all(key in move for key in ("play_type", "card")) and
-            move["play_type"] is "discard")
+    def is_valid(self, game_info):
+        return game_info.can_discard()
 
-class HanabiDiscloseAction(object):
-    def __init__(self, player_id, to_whom, count):
-        self.player_id = player_id
+    def execute(self, table):
+        table.discard_card(self.player_id, self.card)
+
+class HanabiDiscloseAction(HanabiAction):
+    def __init__(self, player_id, to_whom):
+        super(HanabiDiscloseAction, self).__init__(player_id)
         self.to_whom = to_whom
-        self.count = count
+        self.count = 0
+        self.disclosure_type = ''
```

Comment:
Similar idea here are above - it seems to me that nobody would/should ever use a `HanabiDiscloseAction`, so making it abstract and making `disclosure_type` (or a similarly named method) an abstract method seems like it could be a nice way to communicate intent.